### ISSUE TITLE
Multiple binaries for easier usage?

### ISSUE THREAD

**Issue by NoraCodes** [timvisee/ffsend#24]
This may be a personal thing, but I would prefer:

```bash
ffsend [OPTIONS] FILE # Uploads a given file
ffget [OPTIONS] URL # Downloads from a URL
```
to the current interface. Is this something the project would be open to?

**Comment by timvisee** [comment]
I do like the idea, however, I do not want to change the current implementation of having everything in a single binary because this keeps things simple (in the sense that you only have to manage a single binary for distribution, installation and such).

What motivates you to like separate binaries more? Because it better follows the Unix philosophy of having every function contained in a separate binary, to do a single thing well?

I am open for releasing separate binaries as well along with the monolith one, so you have the option of using those instead. I'm not quite sure though how hard it would be to implement that in the current project. That's something I'd have to look into.

**Comment by NoraCodes** [comment]
> And what about a busybox like implementation? Which is still a single binary, with all the separate commands sym-linked to it. Which makes it act like separate programs.

I very much think this is the way to go. `ffsend` for the monolithic binary with `ffget` and `ffput` as symlinks would be great.

> What motivates you to like separate binaries more? Because it better follows the Unix philosophy of having every function contained in a separate binary, to do a single thing well?

More that the concept of most command line tools is "here's what I'm doing, here's what I'm doing it on". When I `ffsend`(or `ffput`) I want to send a file; `ffput myfile.ext`.

**Comment by timvisee** [comment]
@NoraCodes Cool! Because I think such an implementation would be fairly easy, requiring minimal changes.

What about distributing this though? Creating symlinks for every command by default would make packages a little bloated.
Maybe a child package could be used, just to provide the links (this is however not an option for things like `snap`). Or the end user could create the links manually anyway.

**Comment by acgissues** [comment]
Don't see it as necessary when you could just create aliases or a simple script in whichever language you prefer to wrap the existing `ffsend`.

**Comment by timvisee** [comment]
@NoraCodes This has been implemented in a6deb9155b7c9fea580cb9570f45730c6b2234a8, which will be released with [`v0.2.35`](https://github.com/timvisee/ffsend/releases/tag/v0.2.35). The update should be available as soon as this pipeline succeeds:
https://gitlab.com/timvisee/ffsend/pipelines/52250079

Please read the following section I've added to the README about this feature:
https://github.com/timvisee/ffsend#binary-for-each-subcommand-ffput-ffget

Currently, I've only defined [`ffput`, `ffget` and `ffdel`].
If you're interested in more binary names to infer a subcommand for, you may submit a pull request or open a new issue. The actual list of binary names is defined here:
https://github.com/timvisee/ffsend/blob/0c3ac720a57c96d0b1314322ba91032668e2bac2/src/config.rs#L29-L35

I do not have plans right now to implement these links in any of the available packages.

---

@acgissues Using aliases requires a shell, creating wrapping scripts for this is an option but I don't like that too much. Implementing this was actually very simple (see a6deb9155b7c9fea580cb9570f45730c6b2234a8), and the implementation is easy to maintain.

I actually wanted to make the `upload` and `download` subcommands optional, to allow users to upload using a shorter syntax such as `ffsend my-file.txt`. It's hard to properly implement this though, with quite some edge cases. Therefore I abandoned the idea. The approach I've just implemented fits much better I believe.
